### PR TITLE
Swagger 2.0

### PR DESCRIPTION
**Pull request by bezhermoso**
https://github.com/nelmio/NelmioApiDocBundle/pull/705

### CODE REVIEW

**Comment by aminin** [comment]
:+1:

**Comment by SBkubric** [comment]
:+1: @willdurand  Could you please merge this branch with nelmio:master?

**Comment by aminin** [comment]
@SBkubric WIP means Work-In-Progress. Looks like this code was never run. It has no tests, docs yet has some obvious bugs.

File: Tests/Fixtures/app/config/default.yml [hunk 1]
```diff
@@ -75,3 +75,22 @@ nelmio_api_doc:
...
+    swagger2:
+        base_path: /api
+        schemes: ['http', 'https']
+        consumes: ['application/json', 'text/xml']
+        produces: ['application/json', 'text/xml']
+        info:
+          version: 3.14
+          title: Nelmio Swagger2
+          description: Testing Swagger2 integration.
+          termOfService: https://github.com/nelmio/NelmioApiDocBundle/blob/master/README.md
```
Comment (OskarStark) [inline Tests/Fixtures/app/config/default.yml]: i would like ``termsOfService``

File: Tests/Formatter/Swagger2FormatterTest.php [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        parent::setUp();
+
+        $container       = $this->getContainer();
+        $this->extractor = $container->get('nelmio_api_doc.extractor.api_doc_extractor');
+        $this->formatter = $container->get('nelmio_api_doc.formatter.swagger2_formatter');
+    }
+
+    public function testResourceListing()
+    {
+
```
Comment (OskarStark) [inline Tests/Formatter/Swagger2FormatterTest.php]: remove the blank line

File: Tests/Swagger2/ModelRegistryTest.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+        $container       = $this->getContainer();
+        $this->extractor = $container->get('nelmio_api_doc.extractor.api_doc_extractor');
+    }
+
+    public  function testRegister()
+    {
+        $registry = new ModelRegistry();
+
+        $params = array(
```
Comment (OskarStark) [inline Tests/Swagger2/ModelRegistryTest.php]: no assertion in this test?

File: Tests/Formatter/Swagger2FormatterTest.php [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        set_error_handler(array($this, 'handleDeprecation'));
+        $data = $this->extractor->all();
+        restore_error_handler();
+
+        /** @var $formatter SwaggerFormatter */
```
Comment (OskarStark) [inline Tests/Formatter/Swagger2FormatterTest.php]: where is this doc used?

File: Swagger2/SegmentInterface.php [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace Nelmio\ApiDocBundle\Swagger2;
+
+interface SegmentInterface
+{
+    public function toArray();
```
Comment (OskarStark) [inline Swagger2/SegmentInterface.php]: should have a phpDoc bloc

File: Swagger2/Segment/Schema.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    protected $properties = array();
+
+    public function __construct($name)
+    {
+        $this->name = $name;
+    }
+
+    public function toArray()
+    {
+
```
Comment (OskarStark) [inline Swagger2/Segment/Schema.php]: remove blank line

File: Swagger2/Segment/Schema.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+namespace Nelmio\ApiDocBundle\Swagger2\Segment;
+
+use Nelmio\ApiDocBundle\Swagger2\SegmentInterface;
+
+class Schema implements SegmentInterface
+{
+    protected $name;
+
+    protected $properties = array();
```
Comment (OskarStark) [inline Swagger2/Segment/Schema.php]: phpdoc blocks for properties needed

File: Swagger2/Segment/Parameter/SchemaProperty.php [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public function __construct($name)
+    {
+        parent::__construct($name);
+    }
+
+    public function toArray()
+    {
+        $output = parent::toArray();
+        unset($output["required"]);
```
Comment (OskarStark) [inline Swagger2/Segment/Parameter/SchemaProperty.php]: would use single quotes

File: Swagger2/Segment/Parameter/Query.php [hunk 9]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Nelmio\ApiDocBundle\Swagger2\Segment\Parameter;
+
+class Query extends AbstractParameter
+{
+    protected $in = 'query';
```
Comment (OskarStark) [inline Swagger2/Segment/Parameter/Query.php]: phpdoc blocks for properties needed

and should ``$in`` be ``$input`` ?

File: Swagger2/Segment/Parameter/Body.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Nelmio\ApiDocBundle\Swagger2\Segment\Parameter;
+
+use Nelmio\ApiDocBundle\Swagger2\Segment;
+
+class Body extends AbstractParameter
+{
+    protected $in = 'body';
```
Comment (OskarStark) [inline Swagger2/Segment/Parameter/Body.php]: phpdoc blocks for properties needed

File: Swagger2/Segment/Parameter/AbstractParameter.php [hunk 11]
```diff
@@ -0,0 +1,156 @@
+<?php
+
+namespace Nelmio\ApiDocBundle\Swagger2\Segment\Parameter;
+
+use Nelmio\ApiDocBundle\Swagger2\SegmentInterface;
+
+abstract class AbstractParameter implements SegmentInterface
+{
+    protected $name;
```
Comment (OskarStark) [inline Swagger2/Segment/Parameter/AbstractParameter.php]: phpdoc blocks for properties needed

File: Swagger2/SchemaRegistry.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace Nelmio\ApiDocBundle\Swagger2;
+
+use Nelmio\ApiDocBundle\Swagger2\Segment\Schema;
+
+class SchemaRegistry
+{
+    /**
+     * @var array|Schema[]
```
Comment (OskarStark) [inline Swagger2/SchemaRegistry.php]: this looks wrong to me, IMO there is ``Schema[]`` any time

File: Swagger2/ExpandedDefinition.php [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                    if (!isset($parameters[$paramName])) {
+                        $parameters[$paramName] = $pathParameter;
+                    } else {
+                        $parameters[$paramName] = array_merge($parameters[$paramName], $pathParameter);
+                    }
+                }
+            }
+
+            if (count($parameters)) {
+                $data["parameters"] = array_values($parameters);
```
Comment (OskarStark) [inline Swagger2/ExpandedDefinition.php]: would use single quotes

File: Swagger2/ExpandedDefinition.php [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+    public function addPath(Path $path)
+    {
+        $url = $path->getUrl();
+        if (!isset($this->paths[$url])) {
+            $this->paths[$url] = array();
+        }
+        $this->paths[$url][] = $path;
+    }
+
+    private function getPaths()
```
Comment (OskarStark) [inline Swagger2/ExpandedDefinition.php]: would make this protected

File: Swagger2/ExpandedDefinition.php [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+     */
+    private $info;
+    /**
+     * @var array
+     */
+    private $schemes;
+    /**
+     * @var array
+     */
+    private $consumes;
```
Comment (OskarStark) [inline Swagger2/ExpandedDefinition.php]: please add a new line after each property

File: Formatter/Swagger2Formatter.php [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+    public function setBasePath($basePath)
+    {
+        $this->basePath = $basePath;
+    }
+
+    /**
+     * Strips the base path from a URL path.
+     *
+     * @param $path
+     * @return mixed
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: why mixed? looks like the return value is a string in any case, isn't it?

File: Formatter/Swagger2Formatter.php [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+     */
+    public function setBasePath($basePath)
+    {
+        $this->basePath = $basePath;
+    }
+
+    /**
+     * Strips the base path from a URL path.
+     *
+     * @param $path
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: which type is ``$path``

File: Formatter/Swagger2Formatter.php [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+    /**
+     * @param array $schemes
+     */
+    public function setSchemes($schemes)
+    {
+        $this->schemes = $schemes;
+    }
+
+    /**
+     * @param array $basePath
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: is array correct?

File: Formatter/Swagger2Formatter.php [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+     * @param ApiDoc $annotation
+     *                           return string|array
+     */
+    public function formatOne(ApiDoc $annotation)
+    {
+        throw new \BadMethodCallException(sprintf('%s does not support formatting a single ApiDoc only.', __CLASS__));
+    }
+
+    /**
+     * @param array $info
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: array is correct?

File: Formatter/Swagger2Formatter.php [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+            }
+
+        }
+    }
+
+    /**
+     * Format documentation data for one route.
+     *
+     * @param ApiDoc $annotation
+     *                           return string|array
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: bad indention and @-sign missing

File: Formatter/Swagger2Formatter.php [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+        );
+
+        $format = null;
+        $type = null;
+        $reference = null;
+        $enum = null;
+        $items = null;
+
+        foreach ($input as $paramName => $paramDefinition) {
+
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: remove blank line please

File: Formatter/Swagger2Formatter.php [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+            $paramDefinition = array_merge($defaults, $paramDefinition);
+
+            if (isset($this->typeMap[$paramDefinition["actualType"]])) {
+                $type = $this->typeMap[$paramDefinition["actualType"]];
+            } else {
+
+            }
+
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: remove blank line please

File: Formatter/Swagger2Formatter.php [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+        $items = null;
+
+        foreach ($input as $paramName => $paramDefinition) {
+
+            $paramDefinition = array_merge($defaults, $paramDefinition);
+
+            if (isset($this->typeMap[$paramDefinition["actualType"]])) {
+                $type = $this->typeMap[$paramDefinition["actualType"]];
+            } else {
+
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: why this else block?

File: Formatter/Swagger2Formatter.php [hunk 24]
```diff
@@ -0,0 +1,309 @@
...
+                            $parameter['children']
+                        )
+                    );
+                    $property->setCollection(true);
+                    break;
+            }
+        }
+    }
+
+    private function handleParameters(ExpandedDefinition $definition, array $input, $type)
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: would make this method protected

File: Formatter/Swagger2Formatter.php [hunk 25]
```diff
@@ -0,0 +1,309 @@
...
+    public function inputIsOneDimensional(array $params)
+    {
+        $withChildren = array_map(function ($param) {
+            return isset($param['children']) && count($param['children']) > 0;
+        }, $params);
+
+        return count($withChildren) > 0;
+    }
+
+    private function registerSchema($identifier, $parameters, $type = "object")
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: would make this method protected

Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: would use single quotes

File: Formatter/Swagger2Formatter.php [hunk 26]
```diff
@@ -0,0 +1,309 @@
...
+                //var_dump($apiDoc);
+                //var_dump($data);
+                //$body = $this->handleParameters($definition, $data['parameters'], $input['paramType']);
+            }
+
+        }
+        exit;
+
+        return $definition->toArray();
+
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: remove blank line

File: Formatter/Swagger2Formatter.php [hunk 27]
```diff
@@ -0,0 +1,309 @@
...
+                        $formParam->setType(TypeMap::type($paramDefinition['actualType']));
+                        $path->addParameter($formParam);
+                    }
+                }
+                var_dump($input);
+                var_dump($data);
+                echo "======\n";
+                echo "======\n";
+
+                //var_dump($apiDoc);
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: please remove these var dumps

File: Formatter/Swagger2Formatter.php [hunk 28]
```diff
@@ -0,0 +1,309 @@
...
+                    $body->setSchema($this->registerSchema($identifier, $data["parameters"]));
+                    $path->addParameter($body);
+                } elseif ($this->inputIsOneDimensional($input['parameters'])) {
+                    foreach ($data["parameters"] as $paramDefinition) {
+                        $formParam = new Segment\Parameter\FormData($name);
+                        $formParam->setType(TypeMap::type($paramDefinition['actualType']));
+                        $path->addParameter($formParam);
+                    }
+                }
+                var_dump($input);
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: ??

File: Formatter/Swagger2Formatter.php [hunk 29]
```diff
@@ -0,0 +1,309 @@
...
+
+                    $queryParameter->setType($type);
+                    $queryParameter->setDescription($filter["description"]);
+                    $path->addParameter($queryParameter);
+                }
+            }
+
+            if (isset($data['parameters'])) {
+
+                $identifier = $input["class"];
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: please use single quotes

File: Formatter/Swagger2Formatter.php [hunk 30]
```diff
@@ -0,0 +1,309 @@
...
+
+                    $filter = array_merge(array(
+                        "dataType" => "string",
+                        "actualType" => "string",
+                        "description" => null,
+                    ), $filter);
+
+                    $queryParameter = new Segment\Parameter\Query($name);
+
+                    $type = TypeMap::type($filter["actualType"], "string");
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: please use single quotes

File: Formatter/Swagger2Formatter.php [hunk 31]
```diff
@@ -0,0 +1,309 @@
...
+            $path->setMethods($route->getMethods());
+            $path->setDescription($apiDoc->getDescription());
+
+            $data = $apiDoc->toArray();
+
+            if (isset($data["filters"])) {
+                foreach ($data["filters"] as $name => $filter) {
+
+                    $filter = array_merge(array(
+                        "dataType" => "string",
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: please use single quotes

File: Formatter/Swagger2Formatter.php [hunk 32]
```diff
@@ -0,0 +1,309 @@
...
+
+            $path = new Segment\Path($url);
+
+            $responses = array();
+
+            foreach ($compiled->getPathVariables() as $paramValue) {
+                $parameter = new Segment\Parameter\Path($paramValue);
+
+                if ($paramValue === "_format" && false != ($req = $route->getRequirement("_format"))) {
+                    $parameter->setEnum(explode("|", $req));
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: please use single quotes

File: Formatter/Swagger2Formatter.php [hunk 33]
```diff
@@ -0,0 +1,309 @@
...
+            $url = $this->stripBasePath($route->getPath());
+
+            $path = new Segment\Path($url);
+
+            $responses = array();
+
+            foreach ($compiled->getPathVariables() as $paramValue) {
+                $parameter = new Segment\Parameter\Path($paramValue);
+
+                if ($paramValue === "_format" && false != ($req = $route->getRequirement("_format"))) {
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: please use single quotes

File: Formatter/Swagger2Formatter.php [hunk 34]
```diff
@@ -0,0 +1,309 @@
...
+            if (!is_array($input)) {
+                $input = array(
+                    'class' => $input,
+                    'paramType' => 'form',
+                );
+            } elseif (empty($input['paramType'])) {
+                $input['paramType'] = 'form';
+            }
+
+
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: remove one blank line

File: Formatter/Swagger2Formatter.php [hunk 35]
```diff
@@ -0,0 +1,309 @@
...
+use Nelmio\ApiDocBundle\Swagger2\Segment;
+
+/**
+ * Class Swagger2Formatter
+ *
+ * @author Bez Hermoso <bezalelhermoso@gmail.com>
+ */
+class Swagger2Formatter implements FormatterInterface
+{
+    protected $info = array();
```
Comment (OskarStark) [inline Formatter/Swagger2Formatter.php]: add phpdoc to properties

File: DependencyInjection/Swagger2ConfigCompilerPass.php [hunk 36]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+/*
```
Comment (OskarStark) [inline DependencyInjection/Swagger2ConfigCompilerPass.php]: i'm not sure if this header is correct

**Comment by OskarStark** [comment]
Status: reviewd

**Comment by joelwurtz** [comment]
May be a good thing to rename this Formatter into OpenApiFormatter (or OAIFormatter), as it's the new name for swagger 2.0 specification according to this : https://openapis.org/

**Comment by willdurand** [comment]
I like @joelwurtz's suggestion

**Comment by joelwurtz** [comment]
Also don't know if this will help but there is here : https://github.com/jolicode/jane-swagger a complete set of class / normalizer to represent a swagger spec 2.0 into a model object and allow serialization / deserialization of this model object (if there is interest in this, i can split this model into another repository which follow the swagger (openapi) specification)

File: Swagger2/SchemaRegistry.php [hunk 37]
```diff
@@ -0,0 +1,71 @@
...
+        $schemaProperties = array();
+
+        if (is_array($parameters)) {
+            foreach ($parameters as $name => $parameter)
+            {
+                if (!isset($parameters['type'])) {
+                    continue;
+                }
+                var_dump($name);
+                var_dump($parameter);
```
Comment (pylebecq) [inline Swagger2/SchemaRegistry.php]: To remove.

File: Swagger2/SchemaRegistry.php [hunk 38]
```diff
@@ -0,0 +1,71 @@
...
+                $property = new Segment\Parameter\SchemaProperty($name);
+                $schemaProperties[] = $property;
+
+                switch ($parameter['actualType']) {
+                case DataTypes::MODEL:
```
Comment (pylebecq) [inline Swagger2/SchemaRegistry.php]: Missing use statement it seems.

**Comment by tristanbes** [comment]
what's the status of this PR ?

**Comment by hpatoio** [comment]
You might be interested in https://blog.liip.ch/archive/2016/05/11/convert-nelmioapidocbundle-to-swagger-php.html

**Comment by willdurand** [comment]
Closing in favor of: https://blog.liip.ch/archive/2016/05/11/convert-nelmioapidocbundle-to-swagger-php.html.

**Comment by tristanbes** [comment]
not sure if closing this @willdurand is a good move since the command you reported only converts the annotation to swagger, but there is no tool able to dump the `swagger.json` file base on this annotation reading. Maybe i'm wrong ?

For example, having this command when you want to use apiary or gelato to have a nice api documentation portal it's useless for this use case (because they require a `swagger.json` file.

**Comment by hpatoio** [comment]
@tristanbes The command in the linked article convert the NelmioApiDocBundle annotations to annotation readable by Swagger-PHP https://github.com/zircote/swagger-php/  Once you ahve converted the annotation Swagger-PHP can then produce a `swagger.json` file. 

The "problem" if you take the way described in the article is that you loose ApiNelmioDocBundle annotations and so you don't have the bundle functionalities any more.

**Comment by manuelbcd** [comment]
It would be really fantastic to consider the chance to maintain the compatibility feature natively.

**Pull request closed without merging**